functionary/prompt.py

```diff
@@ -31,147 +101,191 @@ def from_message(cls, message: Dict) -> EndToken:
+            else:  # if no function call and content is None --> this is used at inference
+                text = "assistant:"
+
+        return text
+
+    def get_stop_tokens_for_generation(self) -> List[str]:
+        return [self.end_assistant, self.end_function_call]
+
+    def get_assistant_prefixes(self) -> List[str]:
+        result = []
+        for item in [self.end_user, self.end_function]:
+            prefix = f"{item}\nassistant:"
+            result.append(prefix)
+        return result
+
+    def parse_assistant_response(self, llm_ouput: str) -> Dict:
+        generated_content = llm_ouput.strip()
+        for endtoken in self.get_stop_tokens_for_generation():
+            if generated_content.endswith(endtoken):
+                generated_content = generated_content[: -len(endtoken)].strip()
+        # First we need to check if llm_output contains start_token or not
+        start_function_index = generated_content.find(self.start_function)
+        text_content = generated_content
+        result = {"role": "assistant", "content": None}
+        if start_function_index >= 0:
+            func_info = generated_content[
+                start_function_index + len(self.start_function) :
+            ].strip()
+            index = func_info.find(":")
+            func_name = func_info[:index].strip()
+            arguments = func_info[index + 1 :].strip()
+            text_content = generated_content[:start_function_index].strip()
+            result["function_call"] = {
+                "name": func_name,
+                "arguments": arguments,
+            }  # FunctionCall(name=func_name, arguments=arguments)
+        if len(text_content) > 0:
+            result["content"] = text_content
+        return result
+
+
+class PromptTemplateV2(PromptTemplate):
+    from_token = "<|from|>"
+    recipient_token = "<|recipient|>"
+    content_token = "<|content|>"
+    stop_token = "<|stop|>"
+
+    def get_additional_tokens(self) -> List[str]:
+        return [
+            self.from_token,
+            self.recipient_token,
+            self.content_token,
+            self.stop_token,
+        ]
+
+    def get_text_from_message(self, message: Dict) -> str:
+        role = message["role"]
+        content = message.get("content", None)
 
-    elif message["role"] == "user" and content is None:
-        text = "user:\n"
+        if role in ["system", "user"]:  # <from>system\n<recipient>all\n<content>xxx
+            return f"{self.from_token}{role}\n{self.recipient_token}all\n{self.content_token}{content}\n"
 
-    elif message["role"] == "user":
-        text = f"user:\n{content}{end_token}\n"
+        if role == "tool":  # <from>tool_name\n<recipient>all\n<content>xxx
+            tool_name = message["name"]
+            return f"{self.from_token}{tool_name}\n{self.recipient_token}all\n{self.content_token}{content}\n"
 
-    elif message["role"] == "assistant":
+        assert role == "assistant"
```

Comment:
We should validate the message for consistency. a function to validate the message, and throw exception if its not passing the validation.
We can do it with taking a ChatMessage and then validating it there maybe? Or somewhere better suited. Maybe in this function's first couple of lines, before doing anything.

Reply:
Actually this is to make sure that role must be one of: [user, system, assistant, tool]